File: Python/ExampleTests/production/Catalog.py
from catalog_item import CatalogItem


class Catalog:
    MINIMUM = 0
    MAXIMUM = 10000000

    def __init__(self):
        self._items: list[CatalogItem] = []

    def add(self, item: CatalogItem) -> None:
        self._items.append(item)

    def delete(self, item: CatalogItem) -> bool:
        try:
            self._items.remove(item)
            return True
        except ValueError:
            return False

    def read(self) -> list[CatalogItem]:
        return list(self._items)

    def update(self, old_item: CatalogItem, new_item: CatalogItem) -> bool:
        try:
            idx = self._items.index(old_item)
            self._items[idx] = new_item
            return True
        except ValueError:
            return False

    def size(self) -> int:
        return len(self._items)

File: TypeScript/ExampleTests/production/Catalog.ts
import { CatalogItem } from "./CatalogItem.js";

export class Catalog {
  static MINIMUM = 0;
  static MAXIMUM = 10000000;

  #items: CatalogItem[] = [];

  add(item: CatalogItem): void { this.#items.push(item); }

  delete(item: CatalogItem): boolean {
    const idx = this.#items.indexOf(item);
    if (idx < 0) return false;
    this.#items.splice(idx, 1);
    return true;
  }

  read(): CatalogItem[] { return [...this.#items]; }

  update(oldItem: CatalogItem, newItem: CatalogItem): boolean {
    const idx = this.#items.indexOf(oldItem);
    if (idx < 0) return false;
    this.#items[idx] = newItem;
    return true;
  }

  size(): number { return this.#items.length; }
}
